src/CombinedRegexp.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Strings;
+
+/**
+ * `CombinedRegexp` optimizes matching of multiple regular expressions.
+ * Read more about the concept in
+ * {@see https://nikic.github.io/2014/02/18/Fast-request-routing-using-regular-expressions.html}.
+ */
+final class CombinedRegexp
+{
+    private const REGEXP_DELIMITER = '/';
+    private const QUOTE_REPLACER = '\\/';
+
+    /**
+     * @var string[]
+     */
+    private array $patterns;
+    private string $compiledPattern;
+
+    /**
+     * @param string[] $patterns Regular expressions to combine.
+     * @param string $flags Regular expression flags.
+     */
+    public function __construct(
+        array $patterns,
+        string $flags = ''
+    ) {
+        if (count($patterns) === 0) {
+            throw new \InvalidArgumentException('At least one pattern should be specified.');
+        }
+        $this->patterns = $patterns;
+        $this->compiledPattern = $this->compilePatterns($patterns) . $flags;
+    }
+
+    /**
+     * @return string The compiled pattern.
+     */
+    public function getCompiledPattern(): string
+    {
+        return $this->compiledPattern;
+    }
+
+    /**
+     * Returns `true` whether the given string matches any of the patterns, `false` - otherwise.
+     */
+    public function matches(string $string): bool
+    {
+        return preg_match($this->compiledPattern, $string) === 1;
+    }
+
+    /**
+     * Returns pattern that matches the given string.
+     * @throws \Exception if the string does not match any of the patterns.
+     */
+    public function matchPattern(string $string): string
```

Comment:
The name could be better. Current one is an action i.e. match the pattern. It could be `getMatchingPattern()` or `matchingPattern()`.